web/src/message_view.js

```diff
@@ -661,6 +661,86 @@ export function show(raw_terms, opts) {
     }
 }
 
+function handle_anchor_case(anchor, has_found_anchor, get_msg_id) {
+    // These functions are scoped inside `handle_anchor_case` to
+    // to avoid them being used for any other purpose.
+    function duplicate_current_msg_list_with_new_data(data) {
+        const msg_list = new message_list.MessageList({data});
+        msg_list.reading_prevented = message_lists.current.reading_prevented;
+        return msg_list;
+    }
+
+    function select_msg_id(msg_id, select_opts) {
+        message_lists.current.select_id(msg_id, {
+            then_scroll: true,
+            from_scroll: false,
+            ...select_opts,
+        });
+    }
+
+    function select_anchor_using_data(data) {
+        const msg_list = duplicate_current_msg_list_with_new_data(data);
+        message_lists.update_current_message_list(msg_list);
+        // `force_rerender` is required to render the new data.
+        select_msg_id(get_msg_id(), {force_rerender: true});
+    }
+
+    if (has_found_anchor(message_lists.current.data.fetch_status)) {
+        select_msg_id(get_msg_id());
+    } else {
+        let msg_list_data = new MessageListData({
+            filter: message_lists.current.data.filter,
+            excludes_muted_topics: message_lists.current.data.excludes_muted_topics,
+        });
+        load_local_messages(msg_list_data);
```

Comment:
I think this can be moved into the first `if` statement unless https://github.com/zulip/zulip/pull/30346/commits/218f4ddc4ab4380d40f0e2cd3647c295be3a1a1c#r1635403677 changes how this works, since we're not actually using this `msg_list_data` object at all in the second block;

Reply:
Yeah, updated to do that.